### PR TITLE
Capture only used values and convert symbols to strings.

### PR DESCRIPTION
**Pull request by Szymon20000**
## Why
Some people like to capture huge objects such as `Animated`. These objects may contain symbols that weren't handled in the past. The problem with symbols can be partially solved by converting symbols to strings. (it's better than nothing). 
Symbols cannot be converted to symbols and it can be seen in the following example: 
A->C, B->C, and C is a symbol. Then  after conversion A->C != B->C.  Another reason to merge the pr is to speed up Reanimated. Converting objects can take a lot of time. By capturing used values only we significantly reduce this time.

## How
I've added `ClosureGenerator` class that's responsible for capturing used paths only. It takes a path with the name of the variable that would be fully captured in the previous implementation and iterates upwards in the babel nodes tree until there is an edge case or the current node's type is different than 'MemberExpression'.  `ClosureGenerator` maintains a Trie tree where all used references are stored. 

### Assigment operator 
If somebody wants to add a new property to the captured value then obviously we don't want to capture it.
a.b.c.d = 5;
captures a.b.c;

### Mutable  
sth.value

### a.b.c[]
Sometimes we can't tell what property the programmer wants to get. In such a case we capture only `a.b.c`

### Methods that use `this`
The problems only occur with objects that have been sent between threads. However, in Reanimated we only allow for Object|Array|String|Number|Boolean|Function types so it's enough to blacklist functions from their prototypes. The blacklist was base on https://developer.mozilla.org/pl/docs/Web/JavaScript/Referencje/Obiekty/Array/prototype.

## Problems
Update2: 
The problem below has been solved by blacklisting Array|Object|String|Number|Boolean prototype methods. It's a correct solution because in Reanimated2 we do not allow other types to be sent between threads.
Update: 
Currently, there is a problem with methods that use 'this'. For Example:
```JS
sth.map(() =>)
```

_____________
How it works:
```JS
function A() {
  'worklet';
  Animated.withSpring(0);
  Animated.a.b.c = 3;
  const x = Animated.a.b;
  const y = Animated.c.d['szymon'];
  const w = Animated.e.r[getPropName()];
  const ee = Animated.c.r[x];
  const ww = Animated.wer.ty[klocek];
}
```
->
```JS
var A = function () {
    var _f = function _f() {
      _Animated.default.withSpring(0);

      _Animated.default.a.b.c = 3;
      var x = _Animated.default.a.b;
      var y = _Animated.default.c.d['szymon'];

      var w = _Animated.default.e.r[getPropName()];

      var ee = _Animated.default.c.r[x];
      var ww = _Animated.default.wer.ty[klocek];
    };

    _f._closure = {
      Animated: {
        withSpring: _Animated.default.withSpring,
        a: {
          b: _Animated.default.a.b
        },
        c: {
          d: _Animated.default.c.d,
          r: _Animated.default.c.r
        },
        e: {
          r: _Animated.default.e.r
        },
        wer: {
          ty: _Animated.default.wer.ty
        }
      },
      getPropName: getPropName,
      klocek: klocek
    };
    _f.asString = "function A(){const{Animated,getPropName,klocek}=jsThis._closure;{Animated.withSpring(0);Animated.a.b.c=3;var x=Animated.a.b;var y=Animated.c.d['szymon'];var w=Animated.e.r[getPropName()];var ee=Animated.c.r[x];var ww=Animated.wer.ty[klocek];}}";
    _f.__workletHash = 12002772052812;

    global.__reanimatedWorkletInit(_f);

    return _f;
  }();
```

fixes: https://github.com/software-mansion/react-native-reanimated/issues/1132
fixes: https://github.com/software-mansion/react-native-reanimated/issues/1051

### CODE REVIEW

**Comment by terrysahaidak** [comment]
I don't see the "szymon" prop in the result example you provided. Is it missing or example is outdated already?

**Comment by Szymon20000** [comment]
Do you mean there is no 'szymon' in closure? That's because there is the [] operator.

**Comment by terrysahaidak** [comment]
Hmm, how does it work with those values? Do you handle them in that pr?

**Comment by Szymon20000** [comment]
Babel adds `computed` flag to a MemberExpression node if it changes [sth] to .sth

**Review by karol-bisztyga — APPROVED** [review]

File: plugin.js [hunk 1]
```diff
@@ -45,6 +47,162 @@ const globals = new Set([
   '_scrollTo',
 ]);
 
+const blacklistedFunctions = new Set([
+  'stopCapturing',
```
Comment (karol-bisztyga) [inline plugin.js:51]: maybe we should leave a note about this

Comment (Szymon20000) [reply plugin.js:51]: What would be a proper place for that?

File: plugin.js [hunk 2]
```diff
@@ -45,6 +47,162 @@ const globals = new Set([
...
+      return notFound;
+    }
+    if (
+      memberExpressionNode.computed ||
+      memberExpressionNode.property.name === 'value' ||
+      blacklistedFunctions.has(memberExpressionNode.property.name)
+    ) {
+      // a.b[w] -> a.b.w in babel nodes
+      // a.v.value
+      // sth.map(() => )
```
Comment (jakub-gonet) [inline plugin.js:131]: is this comment necessary?

File: plugin.js [hunk 3]
```diff
@@ -45,6 +47,162 @@ const globals = new Set([
...
+      return notFound;
+    }
+    if (
+      path.parent &&
+      path.parent.type === 'AssignmentExpression' &&
+      path.parent.left === path.node
+    ) {
+      /// captured.newProp = 5;
```
Comment (jakub-gonet) [inline plugin.js:139]: is this comment necessary?

File: plugin.js [hunk 4]
```diff
@@ -45,6 +47,162 @@ const globals = new Set([
   '_scrollTo',
 ]);
 
+const blacklistedFunctions = new Set([
```
Comment (jakub-gonet) [inline plugin.js:50]: Can you write a comment on how this list of blacklisted was created?

File: plugin.js (hunk 1 above)
Comment (karol-bisztyga) [reply plugin.js:51]: line 49 could be ok, like explaining the constant

File: plugin.js (hunk 2 above)
Comment (Szymon20000) [reply plugin.js:131]: It explains the main edge case in which we don't want to capture the prototypes methods like map for example.

File: plugin.js (hunk 3 above)
Comment (Szymon20000) [reply plugin.js:139]: All those comments are just examples of edge cases.

File: plugin.js (hunk 4 above)
Comment (Szymon20000) [reply plugin.js:50]: Answered in pr description.

**Pull request merged**